File: sol/140.cpp
struct Solution {
    class TrieNode {
        public:
        int idx = -1;
        array<TrieNode *, 26> child = {nullptr};

        ~TrieNode() {
            for (auto &ptr : child) {
                if (ptr != nullptr){
                    delete ptr;
                }
            }
        }
    };

    vector<string> wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        TrieNode root;
        for (int i = 0; i < wordDict.size(); ++i){
            TrieNode *ptr = &root;
            for (char &c : wordDict[i]){
                if (ptr->child[c - 'a'] == nullptr){
                    ptr->child[c - 'a'] = new TrieNode();
                }
                ptr = ptr->child[c - 'a'];
            }
            ptr->idx = i;
        }
        vector<vector<vector<int>>> memo(n);
        if (root.child[s[n - 1] - 'a'] != nullptr && root.child[s[n - 1] - 'a']->idx != -1){
            memo[n - 1].push_back({root.child[s[n - 1] - 'a']->idx});
        }
        for (int i = n - 2; i >= 0; --i){
            TrieNode *ptr = &root;
            bool isBroken = false;
            for (int j = i; j < n - 1; ++j){
                if (ptr->child[s[j] - 'a'] == nullptr){
                    isBroken = true;
                    break;
                }
                ptr = ptr->child[s[j] - 'a'];
                if (ptr->idx != -1){
                    int l = memo[i].size();
                    memo[i].reserve(l + memo[j + 1].size());
                    memo[i].insert(memo[i].end(), memo[j + 1].begin(), memo[j + 1].end());
                    for (int k = 0; k < memo[j + 1].size(); ++k){
                        memo[i][l + k].push_back(ptr->idx);
                    }
                }
            }
            if (!isBroken){
                ptr = ptr->child[s[n - 1] - 'a'];
                if (ptr != nullptr && ptr->idx != -1){
                    memo[i].push_back({ptr->idx});
                }
            }
        }
        vector<string> res;
        for (auto &p : memo[0]){
            int psize = p.size();
            string s = wordDict[p[psize - 1]];
            for (int i = p.size() - 2; i >= 0; --i){
                s += " ";
                s += wordDict[p[i]];
            }
            res.push_back(s);
        }
        return res;
    }
};

File: sol/140/140.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        trie = dict()
        for i, w in enumerate(wordDict):
            ptr = trie
            for c in w:
                if c not in ptr:
                    ptr[c] = dict()
                ptr = ptr[c]
            ptr['='] = i
        n = len(s)
        memo = [list() for _ in range(n)]
        try:
            memo[n - 1].append([wordDict.index(s[-1])])
        except ValueError:
            pass
        for i in range(n - 2, -1, -1):
            ptr = trie
            for j in range(i, n - 1):
                if s[j] not in ptr:
                    break
                ptr = ptr[s[j]]
                if '=' in ptr:
                    for suff in memo[j + 1]:
                        memo[i].append(suff + [ptr['=']])
            else:
                if (idx := ptr.get(s[n - 1], dict()).get('=', None)) is not None:
                    memo[i].append([idx])
        return [' '.join(wordDict[i] for i in p[::-1]) for p in memo[0]]
